scrapy/core/http2/protocol.py

```diff
@@ -0,0 +1,191 @@
+        self._metadata = {
+            'certificate': None,
+            'ip_address': None
+        }
+
+    def _stream_close_cb(self, stream_id: int):
+        """Called when stream is closed completely
+        """
+        self.streams.pop(stream_id, None)
+
+    def _new_stream(self, request: Request):
+        """Instantiates a new Stream object
+        """
+        stream_id = next(self.stream_id_count)
+
+        stream = Stream(
+            stream_id=stream_id,
+            request=request,
+            connection=self.conn,
+            conn_metadata=self._metadata,
+            write_to_transport=self._write_to_transport,
+            cb_close=self._stream_close_cb
+        )
+
+        self.streams[stream.stream_id] = stream
+        return stream
+
+    def _send_pending_requests(self):
+        # TODO: handle MAX_CONCURRENT_STREAMS
+        # Initiate all pending requests
+        while self._pending_request_stream_pool:
+            stream = self._pending_request_stream_pool.popleft()
+            stream.initiate_request()
+
+    def _write_to_transport(self):
+        """ Write data to the underlying transport connection
+        from the HTTP2 connection instance if any
+        """
+        data = self.conn.data_to_send()
+        self.transport.write(data)
+
+        LOGGER.debug("Sent {} bytes to {} via transport".format(len(data), self._metadata['ip_address']))
+
+    def request(self, _request: Request):
+        stream = self._new_stream(_request)
+        d = stream.get_response()
+
+        # If connection is not yet established then add the
+        # stream to pool or initiate request
+        if self.is_connection_made:
+            stream.initiate_request()
+        else:
+            self._pending_request_stream_pool.append(stream)
+
+        return d
+
+    def connectionMade(self):
+        """Called by Twisted when the connection is established. We can start
+        sending some data now: we should open with the connection preamble.
+        """
+        self.destination = self.transport.getPeer()
+        LOGGER.info('Connection made to {}'.format(self.destination))
+
+        self._metadata['certificate'] = Certificate(self.transport.getPeerCertificate())
+        self._metadata['ip_address'] = ipaddress.ip_address(self.destination.host)
+
+        self.conn.initiate_connection()
+        self._write_to_transport()
+        self.is_connection_made = True
+
+    def dataReceived(self, data):
+        events = self.conn.receive_data(data)
+        self._handle_events(events)
+        self._write_to_transport()
+
+    def connectionLost(self, reason=connectionDone):
+        """Called by Twisted when the transport connection is lost.
+        No need to write anything to transport here.
+        """
+        # Pop all streams which were pending and were not yet started
+        for stream_id in list(self.streams):
+            self.streams[stream_id].close(StreamCloseReason.CONNECTION_LOST, reason)
```

Comment:
The names or the arguments seem a little bit confusing to me, in the sense that `close`'s first parameter is `reason`, but here a variable called "reason" is passed as `failure`. Not a big deal, but perhaps it can be improved somehow.

Reply:
Earlier I was going through what `reason` actually can be used for. I realized that `reason` represents the whole connection which works in HTTP/1.x as it has one request per connection. For HTTP/2 I used it just in case when `connectionLost` is called and we have some streams left then we can pass the `reason: Failure` instance to the errback. Any suggestions on this? 

>  but perhaps it can be improved somehow

Since `reason` is instance of `twisted.python.failure.Failure` I made the argument name of `Stream.close()` as `failure` 😁. Changing the argument name of `reason` in `connectionLost` won't be appropriate here ig?